### DISCUSSION TITLE
Own Collections Framework as Module

### DISCUSSION THREAD

**Discussion by sirinath** [activej/activej#181] [Ideas]
Can ActiveJ add own collection framework. ActiveJ seams to use JDK collections but could benefit from own collection framework.

To keep things simple and fast they may not need to implement the Java Collection API.

Some pointers to existing implementations which can be studied and benchmark against:

Queues:
- https://github.com/conversant/disruptor
- https://github.com/JCTools/JCTools
- https://github.com/LMAX-Exchange/disruptor

Maps:
- https://github.com/leventov/Koloboke
- https://github.com/TimeAndSpaceIO/SmoothieMap

Trees:
- https://github.com/npgall/concurrent-trees

Weak:
- https://github.com/raphw/weak-lock-free

Primitive Specialised:
- http://fastutil.di.unimi.it/
- https://www.eclipse.org/collections/
- https://labs.carrotsearch.com/hppc.html

Misc:
- https://github.com/real-logic/agrona

**Comment by eduard-vasinskyi** [comment]
Hello, @sirinath 
You can use ActiveJ in combination with any collection framework that you like.
We use Java's collections in ActiveJ's components whenever their performance is not a bottleneck. We also use lots of optimized specialized collections (e.g. [ByteBufConcurrentQueue](https://github.com/activej/activej/blob/master/core-bytebuf/src/main/java/io/activej/bytebuf/ByteBufConcurrentQueue.java), [IntrusiveLinkedList](https://github.com/activej/activej/blob/master/util-common/src/main/java/io/activej/common/collection/IntrusiveLinkedList.java), [CaseInsensitiveTokenMap](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/CaseInsensitiveTokenMap.java), etc) as well as external collections in performance-critical parts of ActiveJ code.
For our production projects, we usually use ActiveJ in conjunction with an excellent [HPPC](https://github.com/carrotsearch/hppc) library (see [RingBuffer](https://github.com/activej/activej/blob/master/cloud-memcache/src/main/java/io/activej/memcache/server/RingBuffer.java) in [memcache](https://github.com/activej/activej/tree/master/cloud-memcache/src/main/java/io/activej/memcache) module). External collections can also be serialized by ActiveJ’s [serializer](https://github.com/activej/activej/tree/master/core-serializer/src/test/java/io/activej/serializer) module (see [SerializerDefHppc7Example](https://github.com/activej/activej/blob/master/core-serializer/src/test/java/io/activej/serializer/examples/SerializerDefHppc7Example.java)).

**Comment by sirinath** [comment]
Maybe all the collections can be moved into a separate project. When one gets the bandwidth to create an improved collection over that is in HPPC and other collection frameworks they can be gradually added to the library. Some of the changes can be up-streamed also if they are based on another available collection library.

**Comment by eduard-vasinskyi** [comment]
All of the given collections are highly specialized, they are not general-purpose collections. So it makes no sense to move them outside the module they are being used in. As for collections from other libraries, we use them as-is, without modification. We currently don’t see a reason for creating our own general-purpose collections, given that there are a lot of good external collections that you can use with ActiveJ based on your likings.
